designdocs/CAPACITY.md

```diff
@@ -0,0 +1,404 @@
+
+```rust
+pub enum Event<T: Config> {
+  /// A token account has staked to an MSA.
+  Staked { 
+    /// The token account that staked to an MSA.
+    account: T::AccountId,
+  
+    /// The MSA that a token account staked too.
+    msa_id: MessageSourceId,
+
+    /// An amount that was staked.
+    amount: BalanceOf<T>
+  },
+
+  /// A token account has unstaked to an MSA.
+  UnStake{
+    /// The token account that staked to an MSA.
+    account: T::AccountId,
+
+    /// The MSA that a token account unstaked too.
+    msa_id: MessageSourceId,
+
+    /// An amount that was unstaked.
+    amount: BalanceOf<T>
+   },
+}
+```
+
+Errors when staking
+
+```rust
+pub enum Error<T> {
+  /// Staker tried to stake to the same account more than once.
+  AlreadyStaked,
+
+  /// Staker does not have sufficient balance to cover the amount wanting to stake.
+  InsufficientBalance,
+
+  /// Staker attempted to stake to an invalid MSA.
+  InvalidMsa,
+}
+```
+
+As mentioned above, Capacity is non-transferable and at a implements the following interface to reduce and increase capacity on an MSA.
+
+```rust
+traits Nontransferable {
+   type Balance = Balance;
+
+   /// The available Capacity for an MSA account. 
+   fn available(msa_id: MessageSourceId) -> Result<Balance, DispatchError>;
+   
+   /// Reduce the available Capacity of an MSA account. 
+   fn reduce_available(msa_id: MessageSourceId, amount: Balance) -> Result<Balance, DispatchError>;
+
+   /// Increase the available Capacity for an MSA account.
+   fn increase_available(msa_id: MessageSourceId, amount: Balance) -> Result<Balance, DispatchError>;
+ }
+
+```
+
+**Capacity is replenishable**
+
+Replenishable means that after a certain fixed period, called an epoch, all capacity is replenished. The epoch period is composed of blocks.  Moreover, one unit of Capacity is `10^12` and matches with substrate weight measurement where `10^12` is 1 second of execution time.
+
+Since we want the ability to replenish Capacity the following interface is implement to replenish capacity.
+
+```rust
+trait Replenishable {
+  type Balance: Balance;
+   
+  /// Replenish an MSA amount.
+  fn replenish_by_account(msa_id: MessageSourceId, amount: Balance) -> Result<Balance, DispatchError> {};
+
+  /// Replish all available balance.
+  fn replenish_all_for_account(msa_id: MessageSourceId) -> Result<Balance, DispatchError>;
+
+  /// Checks if an account is able to be replenished.
+  fn can_replenish(msa_id: MessageSourceId) -> bool;
+}
+```
+
+**How capacity replenishes**
+
+As mentioned above, epoch periods are composed of blocks. The next epoch period calculated is based on the current epoch fullness, and expands and contracts depending on a threshold. This threshold is configurable and can be called `config::epochThreshold`. Additionally, we can also configure a multiplier that we increase or decrease the next epoch based on the fullness of the current epoch. This is defined as `config::epochUtilizationMultiplier`.
```

Comment:
We need to shift the dynamic section into the alternatives area.

Reply:
I believe this dynamism mechanism (looks like a simple adaptive window) needs to be detailed further, unless considering V1 of frequency this could fly but not sure adaptive nature takes into consideration number of providers or capacity holders etc.